torch/lib/c10d/ProcessGroup.cpp

```diff
@@ -145,31 +146,25 @@ void ProcessGroup::Work::abort() {
 }
 
 c10::intrusive_ptr<c10::ivalue::Future> ProcessGroup::Work::getFuture() {
-  TORCH_CHECK(false, "ProcessGroup::Work::getFuture not implemented.")
+  return future_;
+}
+
+void ProcessGroup::Work::finish(c10::IValue value) {
+  future_->markCompleted(value);
 }
 
 void ProcessGroup::Work::finish(std::exception_ptr exception) {
-  std::unique_lock<std::mutex> lock(mutex_);
-  completed_ = true;
-  exception_ = exception;
-  if (recordFunctionEndCallback_) {
-    recordFunctionEndCallback_();
-    recordFunctionEndCallback_ = nullptr;
+  if (exception) {
+    future_->setError(exception);
+  } else {
+    future_->markCompleted(c10::IValue());
```

Comment:
This seems odd: we've previously declared the type of the future to be a list of tensors whereas here we're setting the result to a value of "void". Do we need this? Can't we just assume that if the caller is passing us an exception pointer then it means an exception actually occurred (and the pointer isn't null)?

Reply:
This was mostly done for convenience, but we could be more explicit about Work API.